e2e-chatbot-app/app.py

```diff
@@ -0,0 +1,182 @@
+    def save_feedback(index):
+        submit_feedback(
+            endpoint=SERVING_ENDPOINT,
+            request_id=request_id,
+            rating=st.session_state[f"feedback_{index}"]
+        )
+    selection = st.feedback("thumbs", key=f"feedback_{i}", on_change=save_feedback, args=[i])
+
+
+# --- Render chat history ---
+for i, element in enumerate(st.session_state.history):
+    element.render(i)
+
+# --- Chat input (must run BEFORE rendering messages) ---
+if prompt := st.chat_input("Ask a question"):
+    # Add user message to chat history
+    user_msg = UserMessage(content=prompt)
+    st.session_state.history.append(user_msg)
+    user_msg.render(len(st.session_state.history) - 1)
+
+    # Placeholder for assistant response
+    placeholder = st.empty()
+
+    message_buffers = OrderedDict()
+    tool_areas = {}
+
+    with placeholder.container():
+        with st.chat_message("assistant"):
+            response_area = st.empty()
+            response_area.markdown("_Thinking..._")
+
+            input_messages = [msg for elem in st.session_state.history for msg in elem.to_input_messages()]
+            request_id_opt = None
+
+            try:
+                for raw_chunk in query_endpoint_stream(
+                        endpoint_name=SERVING_ENDPOINT,
+                        messages=input_messages,
+                        max_tokens=400,
+                        return_traces=ENDPOINT_SUPPORTS_FEEDBACK
+                ):
+                    response_area.empty()  # Clear previous response
+                    chunk = ChatAgentChunk.model_validate(raw_chunk)
+                    delta = chunk.delta
+                    message_id = delta.id
+                    role = delta.role
+                    content = delta.content
+                    tool_call_id = delta.tool_call_id
+                    request_id = raw_chunk.get("databricks_output", {}).get("databricks_request_id")
+
+                    if request_id:
+                        request_id_opt = request_id
+
+                    if message_id not in message_buffers:
+                        message_buffers[message_id] = {
+                            "chunks": [],
+                            "render_area": st.empty(),
+                        }
+                    message_buffers[message_id]["chunks"].append(chunk)
+
+                    # Live update - render the current partial message's content
+                    partial_message = reduce_chunks(message_buffers[message_id]["chunks"])
+                    render_area = message_buffers[message_id]["render_area"]
+                    with render_area.container():
+                        render_message(partial_message.model_dump_compat(exclude_none=True))
+
+                # Finalize messages and append to history
+                messages = []
+                for msg_id, msg_info in message_buffers.items():
+                    messages.append(reduce_chunks(msg_info["chunks"]))
```

Comment:
Not required, but if we want this to be performant, we can optimize it so that we dont call reduce too many times. As soon as we can store the partial_message on line 157, and if we see a new messageID, that means the partial_message is the full message and append it to a messages array.

It is a minor optimization so not necessary to implement.